api/envoy/extensions/filters/http/router/v3/router.proto

```diff
@@ -104,6 +104,9 @@ message Router {
     }
   }];
 
+  // If set, override the :scheme header to correspond with the upstream transport protocol.
+  bool use_upstream_scheme = 10;
```

Comment:
I prefer an wrapped bool here with default value false. It allows us to change the default value in the future. (Typically different users have different requirements for that. but I personally think use upstream scheme is better than preserve downstream scheme by default. So, at least use a wrapped bool here.)

Reply:
cc @alyssawilk because I found you change the initial behavior (always use upstream secure) to current implementation.

Reply:
Does this need to be a per-route config?  We already have an HCM level config.core.v3.SchemeHeaderTransformation scheme_header_transformation which we could extend to "use upstream scheme"

If you explicitly need per route and don't want this done on a global basis, I suggest you have a per-route SchemeHeaderTransformation so we have consistent global and per-route config for this header.
/wait

Reply:
In the issue linked, I originally proposed using SchemeHeaderTransformation but was unsure how to pass such config from HCM to the router config. So, instead I prototyped a change in SchemeHeaderTransformation (draft PRs targeting private main branch wtzhang23/envoy#1, wtzhang23/envoy#2) instead defining the new field `use_transport_scheme` using the downstream transport protocol when specified in the HCM config and using the upstream's when specified in the Router config. I had qualms with that approach as the same field maps to two different behaviors so I decided to simply open this PR.

Ideally, I agree with you that it should be defined in HCM w/ optional per-route overrides. Just need to figure out the passing of configs.

Reply:
ah so if all you need is HCM, you could worst case stick it as a boolean in stream info in the HCM, and have the router access that field.  There's probably a more elegant way to do so but if all you need is HCM level, let's back out the API change and figure out the right way to get the data from here to there.

Reply:
I'll take a look. I do believe that having per-route overrides and maybe even per-virtual host may be useful for others in the future but as you said I only need it to apply globally.

Reply:
Made changes to HCM in #34099. Let me know if I should close this PR.